[MRG+1] Fix semi_supervised

musically-ut opened the pull request:
Closes #3550, #5774, #3758, #6727 and deprecates #9192.

Pinging @jnothman, @boechat107, @MechCoder. Am I forgetting someone?

boechat107 committed:
Files for my dev environment with Docker

boechat107 committed:
Fixing label clamping (alpha=0 for hard clamping)

boechat107 committed:
Deprecating alpha, fixing its value to zero

boechat107 committed:
Correct way to deprecate alpha for LabelPropagation

The previous way was breaking the test
sklearn.tests.test_common.test_all_estimators

boechat107 committed:
Detailed info for LabelSpreading's alpha parameter

Based on the original paper.

boechat107 committed:
Minor changes in the deprecation message

boechat107 committed:
Improving "deprecated" doc string and raising DeprecationWarning

boechat107 committed:
Using a local "alpha" in "fit" to deprecate LabelPropagation's alpha

This solution isn't great, but it sets the correct value for alpha
without violating the restrictions imposed by the tests.

boechat107 committed:
Removal of my development files

boechat107 committed:
Using sphinx's "deprecated" tag (jnothman's suggestion)

boechat107 committed:
Deprecation warning: stating that the alpha's value will be ignored

jnothman committed:
Use __init__ with alpha=None

jnothman committed:
Update what's new

jnothman committed:
Try fix RuntimeWarning in test_alpha_deprecation

jnothman committed:
DOC Indent deprecation details

jnothman committed:
DOC wording

jnothman committed:
Update docs

musically-ut committed:
Change to the one true implementation.

musically-ut committed:
Add sanity-checked impl. of Label{Propagation,Spreading}

musically-ut committed:
Raise ValueError if alpha is invalid in LabelSpreading.

musically-ut committed:
Add a normalizing step before clamping to LabelPropagation.

jnothman commented:
could you please summarise what makes this wip: what is unresolved?

On 28 Jun 2017 4:57 pm, "Utkarsh Upadhyay" <notifications@github.com> wrote:

> Closes #3550 <https://github.com/scikit-learn/scikit-learn/issues/3550>,
> #5774 <https://github.com/scikit-learn/scikit-learn/issues/5774>, #3758
> <https://github.com/scikit-learn/scikit-learn/pull/3758>, #6727
> <https://github.com/scikit-learn/scikit-learn/pull/6727> and deprecates
> #9192 <https://github.com/scikit-learn/scikit-learn/pull/9192>.
>
> Pinging @jnothman <https://github.com/jnothman>, @boechat107
> <https://github.com/boechat107>, @MechCoder <https://github.com/mechcoder>.
> Am I forgetting someone?
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scikit-learn/scikit-learn/pull/9239
> Commit Summary
>
>    - Files for my dev environment with Docker
>    - Fixing label clamping (alpha=0 for hard clamping)
>    - Deprecating alpha, fixing its value to zero
>    - Correct way to deprecate alpha for LabelPropagation
>    - Detailed info for LabelSpreading's alpha parameter
>    - Minor changes in the deprecation message
>    - Improving "deprecated" doc string and raising DeprecationWarning
>    - Using a local "alpha" in "fit" to deprecate LabelPropagation's alpha
>    - Removal of my development files
>    - Using sphinx's "deprecated" tag (jnothman's suggestion)
>    - Deprecation warning: stating that the alpha's value will be ignored
>    - Use __init__ with alpha=None
>    - Merge branch 'master' into lpalpha
>    - Update what's new
>    - Merge pull request #2 from jnothman/lpalpha
>    - Merge branch 'master' into issue-5774
>    - Try fix RuntimeWarning in test_alpha_deprecation
>    - DOC Indent deprecation details
>    - DOC wording
>    - Update docs
>    - Change to the one true implementation.
>    - Add sanity-checked impl. of Label{Propagation,Spreading}
>    - Raise ValueError if alpha is invalid in LabelSpreading.
>    - Add a normalizing step before clamping to LabelPropagation.
>
> File Changes
>
>    - *M* doc/modules/label_propagation.rst
>    <https://github.com/scikit-learn/scikit-learn/pull/9239/files#diff-0>
>    (4)
>    - *M* doc/whats_new.rst
>    <https://github.com/scikit-learn/scikit-learn/pull/9239/files#diff-1>
>    (5)
>    - *M* examples/semi_supervised/plot_label_propagation_structure.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9239/files#diff-2>
>    (2)
>    - *M* sklearn/semi_supervised/label_propagation.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9239/files#diff-3>
>    (61)
>    - *M* sklearn/semi_supervised/tests/test_label_propagation.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9239/files#diff-4>
>    (66)
>
> Patch Links:
>
>    - https://github.com/scikit-learn/scikit-learn/pull/9239.patch
>    - https://github.com/scikit-learn/scikit-learn/pull/9239.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9239>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67qW6fQwY3Muzsa3kn9eNbATqrJ8ks5sIflvgaJpZM4OHkEw>
> .
>

musically-ut commented:
Just making Travis happy, I think.

Overall, this is a piecemeal effort to fix semi_supervised module as some other issues will remain.
But I believe it is best to fix these first and then take care of the rest in smaller PRs:

 - Updating the tutorial for semi_supervised learning.
 - Fixing implementation of kNN kernel #8008.
 - Adding a few more warnings, re: convergence (e.g. if `max_iter` is reached) and numerical stability (e.g. if an under/overflow occurs during computations).

I'm fixing the flake8 failures as we speak.

musically-ut committed:
Fix flake8 errors.

musically-ut commented:
The failure on appveyor was unrelated to the current changes.

amueller commented:
can you please merge master? that should get rid of the appveyor error.

musically-ut committed:
Remove duplicate imports.

musically-ut commented:
Sigh. I rebased on top of master and had to fix a couple of merge conflicts and it seems that the I was made the 'Committed-by' author for _all_ the commits rather than only the commits which I had fixed. I hope that is not a problem?

Otherwise, I can try to do a fresh merge `--with-ff`.

However, at this point, I don't really trust my git-fu enough to be able to squash these commits in a coherent way. <_<

ogrisel commented:
It's fine, we are going to squash before merging anyway :) Just make sure that the what's new entry reflect the work by all the people who contributed to this fix.

ogrisel commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,68 @@ def test_predict_proba():
+    # adopting notation from Zhou et al (2004):
+    S = clf._build_graph()
+    Y = np.zeros((len(y), n_classes + 1))
+    Y[np.arange(len(y)), y] = 1
```
`Y[:, y] = 1` should work, no?

musically-ut replied:
Surprisingly, no:

```python
In [18]: Y[:, y] = 1

In [19]: Y
Out[19]:
array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.],
       [ 1.,  1.,  1.]])

In [20]: Y = np.zeros((len(y), 3))

In [21]: Y[np.arange(len(y)), y] = 1

In [22]: Y
Out[22]:
array([[ 0.,  0.,  1.],
       [ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.],
       [ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.],
       [ 1.,  0.,  0.],
       [ 1.,  0.,  0.],
       [ 0.,  0.,  1.]])
```

ogrisel replied:
Indeed. I learn everyday.

musically-ut commented:
Ah, right. The scope of this PR and the contributor list has changed slightly; I'll update the What's New.

ogrisel commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,68 @@ def test_predict_proba():
+                               random_state=0)
+    y[::3] = -1
+    Y = np.zeros((len(y), n_classes + 1))
+    Y[np.arange(len(y)), y] = 1
```
same remark

ogrisel commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,68 @@ def test_predict_proba():
+    expected[unlabelled_idx, :] = Y_u
+    expected /= expected.sum(axis=1)[:, np.newaxis]
+
+    assert_array_almost_equal(expected, clf.label_distributions_, 4)
```
If this test take from "Zhou et al (2004)" as well? If so please add a comment.

musically-ut replied:
Not _really_. Joel came up with this "sanity check" by calculating the same value in an iterative way (which is supposed to be numerically stable and should work for large datasets) and the closed form way (which involves inversion of a matrix). The test just compares the two set of values calculated. 

Nevertheless, the closed form formula is indeed from Zhou et al (2004) and it may help any future maintainer. I'll add it.

ogrisel replied:
Thanks.

ogrisel reviewed (changes requested):
Please add the references mentioned at the end of #6727 to the documentation (both the docstring of the classes and the narrative doc if they are missing).

ogrisel commented:
Please also create an issue to track the directed graph issue if not it has not been created already.

ogrisel commented:
I have put some comments but they appear above in the timeline: https://github.com/scikit-learn/scikit-learn/pull/9239#pullrequestreview-47135562

musically-ut committed:
DOC Update What's New.

ogrisel commented:
Also please add the original case you reported as a non regression test:

```
X = np.array([[1., 0.], [0., 1.], [1., 2.5]])
y = np.array([0, 1, -1])
mdl = label_propagation.LabelSpreading(kernel='rbf', max_iter=5000)
mdl.fit(samples, labels)

# this should converge quickly:
assert mdl.n_iter_ < 10
assert_array_equal(mdl.predict(X), [0, 1, 1])
```

musically-ut commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -449,6 +490,9 @@ class LabelSpreading(BaseLabelPropagation):
                  max_iter=30, tol=1e-3, n_jobs=1):
 
+        if alpha <= 0.0 or alpha >= 1.0:
+            raise ValueError('alpha must be inside the open interval (0, 1)')
```
👍

musically-ut committed:
Specify alpha's value in the error.

musically-ut committed:
Tidy up tests.

Add a test and add references, where needed.

musically-ut commented:
@ogrisel Issue #8008 tracks the undirected graphs problem. re: the three references at the end of #6727, two of them are being referred to in the paper.

As far as I can tell, the implementation does not benefit directly from the third reference, i.e. the Bengio et. al. (2006). Our understanding of the problem did improve, but that is only ever directly referred to in the comments of the PR, not in the code. Besides, the paper extends the semi-supervised setting to the inductive setting, which is outside the scope of the current implementation.

musically-ut commented:
The travis failure appears to be incidental because of temporary network issue. I cannot trigger a manual rebuild, but the next commit ought to fix it.

ogrisel commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,80 @@ def test_predict_proba():
+    assert_raises(ValueError, label_propagation.LabelSpreading, alpha=1.1)
+
+
+def test_clamping():
```
Maybe add an inline comment to state that this is a non-regression test.

musically-ut replied:
```python
# Performance test, not a regression test
```
?

ogrisel replied:
Just add a comment such as:

```python
def test_clamping_toy():
    # This is a non-regression test for #5774
    ...
```

ogrisel commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,80 @@ def test_predict_proba():
+                               random_state=0)
+    y[::3] = -1
+    clf = label_propagation.LabelSpreading().fit(X, y)
+    # adopting notation from Zhou et al (2004):
```
2002, right?

musically-ut replied:
`LabelPropagation`'s reference:

> Xiaojin Zhu and Zoubin Ghahramani. Learning from labeled and unlabeled data with label propagation. Technical Report CMU-CALD-02-107, Carnegie Mellon University, 2002 http://pages.cs.wisc.edu/~jerryzhu/pub/CMU-CALD-02-107.pdf

`LabelSpreading`'s reference:

> Dengyong Zhou, Olivier Bousquet, Thomas Navin Lal, Jason Weston, Bernhard Schoelkopf. Learning with local and global consistency (2004) http://citeseer.ist.psu.edu/viewdoc/summary?doi=10.1.1.115.3219

ogrisel replied:
Ah ok sorry.

musically-ut replied:
The names had me confused more than once as well. :)

musically-ut committed:
Add comment to non-regression test.

ogrisel reviewed (approved):
LGTM once my last comment on the non-regression test comment is addressed.

musically-ut commented:
@ogrisel I've added the comment to the test. Thanks for the very thorough review!

ogrisel commented:
Thanks for your work @musically-ut !

@jnothman @amueller shall we merge?

ogrisel commented:
The appveyor failure is caused by a random `HTTPError` on mldata.org. It should be ignored.

jnothman commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +63,81 @@ def test_predict_proba():
+    Y = np.zeros((len(y), n_classes + 1))
+    Y[np.arange(len(y)), y] = 1
+    Y = Y[:, :-1]
+    for alpha in [0.1, 0.3, 0.5, 0.7, 0.9]:
```
I assume this is fast enough, but we could probably remove some alphas if we wish.

musically-ut replied:
I'd rather leave them there; these were giving me some hints as to how to test the numerical stability of the inversion operation in the closed form.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -449,6 +490,10 @@ class LabelSpreading(BaseLabelPropagation):
     def __init__(self, kernel='rbf', gamma=20, n_neighbors=7, alpha=0.2,
                  max_iter=30, tol=1e-3, n_jobs=1):
 
+        if alpha <= 0.0 or alpha >= 1.0:
```
This should happen in `fit`, not `__init__`

jnothman replied:
And I think we need to explicitly forbid `alpha=None`.

musically-ut replied:
In `fit` because of convention?

Otherwise, I would argue that failing as early as possible (i.e. in `__init__`) would be better.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -17,8 +17,8 @@
   The algorithm tries to learn distributions of labels over the dataset. In the
   "Hard Clamp" mode, the true ground labels are never allowed to change. They
```
This comment is no longer really relevant. LabelPropagation requires hard clamping; LabelSpreading requires soft clamping.

musically-ut replied:
I've made the description a bit more generic and about _variants_. `LabelPropagation` and `LabelSpreading` are examples from these different variants.

jnothman reviewed (approved):
I've not looked through this super-thoroughly, but with these changes it should LGTM.

I'm still wondering if we should offer hard clamping for the label spreading graph, and soft clamping for label propagation, i.e. don't deprecate `alpha`. But now that they are distinguished by more than just alpha it might be hard to define correct behaviour. Better to start with something that works!

jnothman commented:
Yes, we do it in fit by convention. Mostly because the value can be changed
by set_params, and we don't want to do a whole lot of descriptor/property
magic.


On 1 July 2017 at 21:20, Utkarsh Upadhyay <notifications@github.com> wrote:

> *@musically-ut* commented on this pull request.
> ------------------------------
>
> In sklearn/semi_supervised/label_propagation.py
> <https://github.com/scikit-learn/scikit-learn/pull/9239#discussion_r125160674>
> :
>
> > @@ -449,6 +490,10 @@ class LabelSpreading(BaseLabelPropagation):
>      def __init__(self, kernel='rbf', gamma=20, n_neighbors=7, alpha=0.2,
>                   max_iter=30, tol=1e-3, n_jobs=1):
>
> +        if alpha <= 0.0 or alpha >= 1.0:
>
> In fit because of convention?
>
> Otherwise, I would argue that failing as early as possible (i.e. in
> __init__) would be better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9239#discussion_r125160674>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wnEMB99GwTOyJrE5Oqc0xIvvW_qks5sJit7gaJpZM4OHkEw>
> .
>

musically-ut committed:
Fix documentation.

musically-ut committed:
Move check for alpha into fit from __init__.

musically-ut commented:
@jnothman I've addressed the changes. Awaiting review.

musically-ut commented:
I don't like the odd failure case when `alpha = None` is actually allowed for `LabelSpreading`. I propose adding another parameter to the `__init__` of the `BaseLabelPropagation` class called `variant`, which will either hold the string `"LabelSpreading"` or `"LabelPropagation"` depending on which sub-class creates it. Then the behavior of clamping will depend on the `variant` variable being equal to one of the strings instead of on whether `alpha is None` or not.

Then we will be able to eliminate the corner case of using `LabelSpreading` and setting `alpha = None`.

Thoughts? 

Any opinions on using strings `LabelSpreading`/`LabelPropagation` or some other way of identifying the variants?

jnothman commented:
You can just do something like:

class LabelSpreading:
    variant = 'spreading'


class LabelPropagation:
    variant = 'propagation'


which would mean you don't need to fuss with __init__

On 2 July 2017 at 03:05, Utkarsh Upadhyay <notifications@github.com> wrote:

> I don't like the odd failure case when alpha = None is actually allowed
> for LabelSpreading. I propose adding another parameter to the __init__ of
> the BaseLabelPropagation class called variant, which will either hold the
> string "LabelSpreading" or "LabelPropagation" depending on which
> sub-class creates it. Then the behavior of clamping will depend on the
> variant variable being equal to one of the strings instead of on whether alpha
> is None or not.
>
> Then we will be able to eliminate the corner case of using LabelSpreading
> and setting alpha = None.
>
> Thoughts?
>
> Any opinions on using strings LabelSpreading/LabelPropagation or some
> other way of identifying the variants?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9239#issuecomment-312443695>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62na7RABCVczzJ_CG-dpmoKyXOEkks5sJnxkgaJpZM4OHkEw>
> .
>

jnothman commented:
And yes, I think this would be a fine way to go.

On 2 July 2017 at 10:12, Joel Nothman <joel.nothman@gmail.com> wrote:

> You can just do something like:
>
> class LabelSpreading:
>     variant = 'spreading'
>
>
> class LabelPropagation:
>     variant = 'propagation'
>
>
> which would mean you don't need to fuss with __init__
>
> On 2 July 2017 at 03:05, Utkarsh Upadhyay <notifications@github.com>
> wrote:
>
>> I don't like the odd failure case when alpha = None is actually allowed
>> for LabelSpreading. I propose adding another parameter to the __init__
>> of the BaseLabelPropagation class called variant, which will either hold
>> the string "LabelSpreading" or "LabelPropagation" depending on which
>> sub-class creates it. Then the behavior of clamping will depend on the
>> variant variable being equal to one of the strings instead of on whether alpha
>> is None or not.
>>
>> Then we will be able to eliminate the corner case of using LabelSpreading
>> and setting alpha = None.
>>
>> Thoughts?
>>
>> Any opinions on using strings LabelSpreading/LabelPropagation or some
>> other way of identifying the variants?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/9239#issuecomment-312443695>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz62na7RABCVczzJ_CG-dpmoKyXOEkks5sJnxkgaJpZM4OHkEw>
>> .
>>
>
>

musically-ut committed:
Fix corner case of LabelSpreading with alpha=None.

musically-ut commented:
Updated; awaiting review.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -264,13 +280,21 @@ def fit(self, X, y):
             self.label_distributions_ = safe_sparse_dot(
                 graph_matrix, self.label_distributions_)
+
+            if alpha == 'propagation':
```
I think this should be self.variant

musically-ut replied:
Gah!

I wish we had a test to see if normalization was working. :(

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -264,13 +280,21 @@ def fit(self, X, y):
                 graph_matrix, self.label_distributions_)
+
+            if alpha == 'propagation':
+                # LabelPropagation
```
and i think this comment is then redundant

musically-ut committed:
alpha -> self.variant

jnothman reviewed (commented):
So is there some way to invent a graph where you get very different normalisation factors for different rows, so that that substantially affects the propagation?

jnothman commented on doc/whats_new.rst:
```diff
@@ -441,7 +441,16 @@ Bug fixes
+   - Fix :class:`semi_supervised.BaseLabelPropagation` to correctly implement
+     ``LabelPropagation`` and ``LabelSpreading`` as done in the referenced
+     papers. :class:`semi_supervised.LabelPropagation` now always does hard
+     clamping. Its ``alpha`` parameter now defaults to ``None`` and is
```
I think you should say that "alpha has no effect and is deprecated to be ...".

musically-ut replied:
👍

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -239,20 +241,34 @@ def fit(self, X, y):
+        if self.variant == 'propagation':
+            # LabelPropagation
+            clamp_weights = np.ones((n_samples, 1))
+            clamp_weights[~unlabeled, 0] = 0.0
```
This should be the same as `clamp_weights = unlabelled.reshape(-1, 1)`

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -239,20 +241,34 @@ def fit(self, X, y):
+            clamp_weights[~unlabeled, 0] = 0.0
+        else:
+            # LabelSpreading
+            clamp_weights = alpha * np.ones((n_samples, 1))
```
just use `clamp_weights = alpha`

musically-ut replied:
Have removed `clamp_weights` completely in favor of keeping only `alpha`.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -264,13 +280,20 @@ def fit(self, X, y):
+
             # clamp
             self.label_distributions_ = np.multiply(
                 clamp_weights, self.label_distributions_) + y_static
```
Hmm. In the label prop case, we could just be using `np.where(unlabelled, self.label_distributions, y_static)` unless I'm mistaken

musically-ut replied:
Yes, `np.where(unlabeled[:, np.newaxis], self.label_distributions_, y_static)` to be more precise. Have factored out `unlabeled[:, np.newaxis]` in the actual implementation.

musically-ut committed:
Make Whats_new more explicit.

musically-ut committed:
Simplify impl. of Label{Propagation,Spreading}.

musically-ut commented:
Err ... the Travis failures, this time, seem to be unrelated to this PR.

ogrisel commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -350,6 +373,14 @@ class LabelPropagation(BaseLabelPropagation):
     LabelSpreading : Alternate label propagation strategy more robust to noise
     """
 
+    variant = 'propagation'
```
Please make this a private attribute `_variant`. Otherwise it could be consider public API.

@jnothman I think we can merge once this is made private. Ok on your side?

ogrisel commented:
You can ignore tests failing with "HTTP Error 500" on mldata.org.

musically-ut committed:
variant -> _variant.

musically-ut commented:
Updated.

PS: I don't remember whether merely pushing commits sends out an notification e-mail to subscribers by default or not on GitHub (I vaguely recall turning off those notifications for myself somewhere). In case it does, please tell me and I'm sorry about the noise coming from _these_ "Updated" comments.

ogrisel commented:
It's fine, let's wait for @jnothman's feedback and CI results.

jnothman commented:
Yes, I'm +1 for merge, because it would be good to finally have this fixed in the release. But it would be nice if we had a non-regression test for that normalisation issue. Want to think about it in the background and try propose something, @musically-ut?

ogrisel commented:
I am will be offline soon. @jnothman I'll you handle the end and merging of this PR with @musically-ut :)

musically-ut commented:
👍  

I am going to continue proposing some changes after the underlying algorithm is fixed (e.g. #8008), and will keep trying to come up with better tests in the background.

@ogrisel Thanks for the review!

jnothman commented:
Thanks and well done, @musically-ut, @boechat107, @MechCoder and all.

amueller commented:
AWESOME! thank you so much to everybody working on this!
